### ISSUE TITLE
Shuffling of IterableDataset disabled when dataloader is passed to accelerator.prepare()

### ISSUE THREAD

**Issue by loubnabnl** [huggingface/accelerate#375]
Hi, I have an issue with shuffling an `IterableDataset` that is passed to a dataloader in `accelerate`. I am trying to use `ShufflerIterDataPipe` from `torch.utils.data.datapipes.iter.combinatorics`.
As you can see below, it looks like the shuffling is disabled:

```python
from accelerate import Accelerator
from datasets import load_dataset
from torch.utils.data.dataloader import DataLoader
from torch.utils.data.datapipes.iter.combinatorics import ShufflerIterDataPipe

data = load_dataset("lvwerra/codeparrot-clean-valid", streaming=True, split="train")
shuffled_data = ShufflerIterDataPipe(data, buffer_size=100)
dataloader = DataLoader(shuffled_data, batch_size=4, shuffle=True)
print(f"Before accelerate prepare: {dataloader.dataset._shuffle_enabled}")

accelerator = Accelerator()
dataloader = accelerator.prepare(dataloader)
print(f"After accelerate prepare: {dataloader.dataset._shuffle_enabled}")
```

Output:

```
Before accelerate prepare: True
After accelerate prepare: False
```

I think it’s because `ShufflerIterDataPipe` returns a new Pytorch `IterableDataset` that is also an `IterDataPipe`. When such instance is passed to a Pytorch dataloader, for the shuffling to be enabled we must specify  `shuffle=True`  in the dataloader, the distinction is done [here](https://github.com/pytorch/pytorch/blob/d1bb420fb1087d680b7ad526536e14acbc2f18bf/torch/utils/data/dataloader.py#L260). I don't think this is handled during the dataloader preparation in `accelerate`.

```
Python 3.9.12
torch=1.11.0
accelerate=0.8.0

compute_environment: LOCAL_MACHINE
deepspeed_config: {}
distributed_type: MULTI_GPU
machine_rank: 0
main_process_ip: null
main_process_port: null
main_training_function: main
mixed_precision: fp16
num_machines: 1
num_processes: 2
use_cpu: false
```

**Comment by lvwerra** [comment]
You mean `dispatch_batch`? That's `true` by default for iterable datasets, no?

**Comment by sgugger** [comment]
Ah yes, it is indeed `True` by default. The problem is that PyTorch does not store the `shuffle=True` argument sent to the first `DataLoader`, which makes it then impossible to re-pass it when instantiating the `DataLoaderDispatcher`.

So maybe we could try to look for that attribute on the `dataset` as you are doing in your test, then manually call again `torch.utils.data.graph_settings.apply_shuffle_settings(dataset, shuffle=shuffle)` in the init of `DataLoaderDispatcher`?

**Comment by loubnabnl** [comment]
Yes I'm working on something similar to that

**Comment by loubnabnl** [comment]
It works with `torch.utils.data.graph_settings.apply_shuffle_settings(dataset, shuffle=shuffle)` I can make a PR to add this
